// abuseipdb/dist/__index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _apiReach = _interopRequireDefault(require("api-reach"));

var _bottomLineUtils = require("bottom-line-utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const {
  REMOVE
} = _bottomLineUtils.mapValues;

class AbuseIPDB {
  /**
   * @param {string} apiKey - AbuseIPDB api key
   */
  constructor(apiKey) {
    this._api = new _apiReach.default({
      base: "https://api.abuseipdb.com/api/v2/",
      headers: {
        Key: apiKey,
        Accept: "application/json"
      }
    });
  }
  /**
   * @param {string} ipAddress - ipv4 or ipv6 to check
   * @param {number} [maxAgeInDays] - how far back in time to fetch reports
   * @param {boolean} [verbose] - if reports array should be returned as well with basic data
   * @returns {Promise}
   */


  check(ipAddress, maxAgeInDays, verbose) {
    const params = (0, _bottomLineUtils.mapValues)({
      ipAddress,
      maxAgeInDays,
      verbose
    }, value => {
      if (!value) {
        return REMOVE;
      }

      return value;
    });
    return this._api.get("/check", params).then(r => r.body.data);
  }

}

var _default = AbuseIPDB;
exports.default = _default;
